File: app/src/main/java/com/soya/launcher/ui/fragment/WallpaperFragment.java
package com.soya.launcher.ui.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.leanback.widget.ArrayObjectAdapter;
import androidx.leanback.widget.FocusHighlight;
import androidx.leanback.widget.FocusHighlightHelper;
import androidx.leanback.widget.ItemBridgeAdapter;
import androidx.leanback.widget.VerticalGridView;

import com.soya.launcher.App;
import com.soya.launcher.R;
import com.soya.launcher.adapter.WallpaperAdapter;
import com.soya.launcher.bean.Wallpaper;
import com.soya.launcher.enums.Atts;
import com.soya.launcher.enums.IntentAction;
import com.soya.launcher.utils.GlideUtils;
import com.soya.launcher.utils.PreferencesUtils;

public class WallpaperFragment extends AbsFragment{

    public static WallpaperFragment newInstance() {

        Bundle args = new Bundle();

        WallpaperFragment fragment = new WallpaperFragment();
        fragment.setArguments(args);
        return fragment;
    }

    private TextView mTitleView;
    private VerticalGridView mContentGrid;
    private ImageView mWallpaperView;

    private ArrayObjectAdapter mArrayObjectAdapter;

    @Override
    public int getLayoutId() {
        return R.layout.fragment_wallpaper;
    }

    @Override
    protected void init(View view, LayoutInflater inflater) {
        super.init(view, inflater);
        mTitleView = view.findViewById(R.id.title);
        mContentGrid = view.findViewById(R.id.content);
        mWallpaperView = view.findViewById(R.id.wallpaper);

        mTitleView.setText(getString(R.string.wallpaper));
    }

    @Override
    protected void initBind(View view, LayoutInflater inflater) {
        super.initBind(view, inflater);
        setContent();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        requestFocus(mContentGrid);
    }

    private void setContent(){
        mArrayObjectAdapter = new ArrayObjectAdapter(new WallpaperAdapter(getActivity(), getLayoutInflater(), newWallpaperCallback()));
        ItemBridgeAdapter itemBridgeAdapter = new ItemBridgeAdapter(mArrayObjectAdapter);
        FocusHighlightHelper.setupBrowseItemFocusHighlight(itemBridgeAdapter, FocusHighlight.ZOOM_FACTOR_MEDIUM, false);
        mContentGrid.setAdapter(itemBridgeAdapter);
        mContentGrid.setNumColumns(3);
        mArrayObjectAdapter.addAll(0, App.WALLPAPERS);
        mContentGrid.requestFocus();
    }

    public WallpaperAdapter.Callback newWallpaperCallback(){
        return new WallpaperAdapter.Callback() {
            @Override
            public void onSelect(boolean select, Wallpaper bean) {
                if (select) GlideUtils.bindBlurCross(getActivity(), mWallpaperView, bean.getPicture(), 800);
            }

            @Override
            public void onClick(Wallpaper bean) {
                PreferencesUtils.setProperty(Atts.WALLPAPER, bean.getId());
                mArrayObjectAdapter.notifyArrayItemRangeChanged(0, mArrayObjectAdapter.size());
                getActivity().sendBroadcast(new Intent(IntentAction.ACTION_UPDATE_WALLPAPER));
            }
        };
    }
}

File: app/src/main/java/com/soya/launcher/ui/fragment/WallpaperFragment.kt
package com.soya.launcher.ui.fragment

import android.os.Bundle
import com.drake.brv.utils.addModels
import com.drake.brv.utils.mutable
import com.drake.brv.utils.setup
import com.shudong.lib_base.base.BaseViewModel
import com.shudong.lib_base.ext.UPDATE_WALLPAPER_EVENT
import com.shudong.lib_base.ext.clickNoRepeat
import com.shudong.lib_base.ext.dimenValue
import com.shudong.lib_base.ext.no
import com.shudong.lib_base.ext.padding
import com.shudong.lib_base.ext.sendLiveEventData
import com.soya.launcher.BaseWallPaperFragment
import com.soya.launcher.R
import com.soya.launcher.bean.Wallpaper
import com.soya.launcher.cache.AppCache.WALLPAPERS
import com.soya.launcher.databinding.FragmentWallpaperBinding
import com.soya.launcher.databinding.HolderWallpaperBinding
import com.soya.launcher.enums.Atts
import com.soya.launcher.product.base.product
import com.soya.launcher.utils.GlideUtils
import com.soya.launcher.utils.PreferencesUtils

class WallpaperFragment : BaseWallPaperFragment<FragmentWallpaperBinding, BaseViewModel>() {

    override val shouldSetPaddingTop: Boolean = false

    override fun initView() {
       mBind.apply {
           layout.title.text = getString(R.string.wallpaper)
       }

        val isShowTitle = product.isShowPageTitle()
        isShowTitle.no { mBind.llRoot.padding(topPadding = com.shudong.lib_dimen.R.dimen.qb_px_15.dimenValue()) }
    }

    override fun initdata() {
        setContent()
        mBind.apply {
            content.apply {
                post {
                    requestFocus()
                }
            }
        }
    }


    private fun setContent() {
        mBind.content.setup {
            addType<Wallpaper>(R.layout.holder_wallpaper)
            onBind {
                val dto = getModel<Wallpaper>()
                val binding = getBinding<HolderWallpaperBinding>()
                binding.root.setCallback{
                    if(it)GlideUtils.bindBlurCross( mBind.wallpaper, dto.picture, 800)
                }
                itemView.clickNoRepeat {
                    dto.id.let { PreferencesUtils.setProperty(Atts.WALLPAPER, it) }
                    notifyItemRangeChanged(0,mBind.content.mutable.size)
                    //mArrayObjectAdapter!!.notifyArrayItemRangeChanged(0, mArrayObjectAdapter!!.size())
                    sendLiveEventData(UPDATE_WALLPAPER_EVENT,true)
                }
            }
        }

        mBind.content.setNumColumns(3)
        mBind.content.addModels(WALLPAPERS)
        mBind.content.requestFocus()
    }


    companion object {
        @JvmStatic
        fun newInstance(): WallpaperFragment {
            val args = Bundle()

            val fragment = WallpaperFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
